### PR TITLE
Fixed #29286 -- Fixed column mismatch crash with QuerySet.values() or values_list() after combining an annotated and unannotated queryset with union(), difference(), or intersection().

### PR DESCRIPTION
**Pull request by felixxm**
https://code.djangoproject.com/ticket/29286

Regression in a0c03c62a8ac586e5be5b21393c925afa581efaf.

### CODE REVIEW

**Comment by felixxm** [comment]
@carltongibson Thanks for the review :+1: . I simplified patch by using unpacking generalization.

File: django/db/models/sql/compiler.py [hunk 1]
```diff
@@ -408,9 +408,8 @@ def get_combinator_sql(self, combinator, all):
                 # If the columns list is limited, then all combined queries
                 # must have the same columns list. Set the selects defined on
                 # the query on all combined queries, if not already set.
-                if (not compiler.query.values_select and not compiler.query.annotations and
-                        self.query.values_select):
-                    compiler.query.set_values(self.query.values_select)
+                if not compiler.query.values_select and self.query.values_select:
+                    compiler.query.set_values((*self.query.values_select, *self.query.annotation_select))
```
Comment (charettes) [inline django/db/models/sql/compiler.py]: Do you want `set_values(*self.query.values_select, *self.query.annotation_select)` instead?

Comment (charettes) [reply django/db/models/sql/compiler.py]: Ahh nvm, the function accepts a tuple and not `*args`.

**Comment by carltongibson** [comment]
Slight aside: 

>  I simplified patch by using unpacking generalization.

The trouble with all this single commit/force pushing is that I need to check what I pulled locally to see what you changed. 

(That done, I had to look twice, but, yes, it's nicer. 🙂)

**Comment by fjsj** [comment]
@felixxm I'm facing a similar issue, but only when using Postgres. Unfortunately, looks like this solution isn't enough.

Here's the git patch file that creates new failing test case. Apply it over `b2a428ec66a409b90f32e2c3d9bf1f984d68132d`:
[0001-Additional-broken-test-for-29286.patch.txt](https://github.com/django/django/files/1903790/0001-Additional-broken-test-for-29286.patch.txt)

Here's my `tests/test_postgres.py`:
```
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.postgresql_psycopg2',
        'NAME': 'django-test',
        'HOST': '127.0.0.1',
    },
    'other': {
        'ENGINE': 'django.db.backends.postgresql_psycopg2',
        'NAME': 'django-test-other',
    }
}

SECRET_KEY = "django_tests_secret_key"

# Use a fast hasher to speed up tests.
PASSWORD_HASHERS = [
    'django.contrib.auth.hashers.MD5PasswordHasher',
]
```

And here's how to run it:
```
./runtests.py --settings=tests.test_postgres queries.test_qs_combinators.QuerySetSetOperationTests.test_union_with_values_on_annotated_and_not_annotated_different_models
```

**Comment by felixxm** [comment]
@fjsj Thanks for the report. IMO your use case is unrelated with this PR and it is unrelated with regression introduced in a0c03c62a8ac586e5be5b21393c925afa581efaf. It didn't work correctly before we add `values()/values_list()` support in 2d3cc94284674638c334670903d49565039d77ae, it didn't work before this regression (a0c03c62a8ac586e5be5b21393c925afa581efaf) and it also didn't work before this patch. I think it has never worked.

If you use `values()/values_list()` then annotations are always after models' fields (see [query.py](https://github.com/django/django/blob/master/django/db/models/query.py#L102-L106)):
```sql
SELECT "queries_number"."num", Foo AS "name"
FROM "queries_number"
UNION
SELECT "queries_reservedname"."name", "queries_reservedname"."order"
FROM "queries_reservedname" 
```
It probably cannot be fix without refactoring the  `values()/values_list()` mechanism. Personally I'm +1 to document this limitation.

**Comment by fjsj** [comment]
Thanks @felixxm , so I guess it's better to create a new issue, right?

**Comment by felixxm** [comment]
I think so. If it is a real-life use case and you're aware of this limitation then you can simple fix this qs by reordering a columns lists in `values()` calls:
```python
qs1 = Number.objects.annotate(
    name=Value('Foo', output_field=TextField())
).values('num', 'name')
qs2 = ReservedName.objects.values('order', 'name')
qs1.union(qs2)
```

**Comment by fjsj** [comment]
Thanks @felixxm , unfortunately reordering is not enough for all cases. Added another failing test where reordering doesn't help here: https://code.djangoproject.com/ticket/28553#comment:3

Sorry for hijacking this PR, I thought the problems were related.

**Comment by felixxm** [comment]
np, it's great that you've found the right ticket. I agree that this limitation should be documented.

I rebased from master due to conflicts (sorry @carltongibson :wink: ).

File: docs/releases/1.11.13.txt [hunk 2]
```diff
@@ -15,3 +15,8 @@ Bugfixes
 
 * Fixed crashes in ``django.contrib.admindocs`` when a view is a callable
   object, such as ``django.contrib.syndication.views.Feed`` (:ticket:`29296`).
+
+* Fixed a regression in Django 1.11.12 where ``QuerySet.values()/values_list()``
```
Comment (timgraham) [inline docs/releases/1.11.13.txt]:
```
* Fixed a regression in Django 1.11.12 where ``QuerySet.values()`` or
  ``values_list()`` after combinging an annotated and unannotated queryset with
  ``union()``, ``difference()``, or ``intersection()`` crashed due to mismatching
  columns (:ticket:`29286`).
```

File: tests/queries/test_qs_combinators.py [hunk 3]
```diff
@@ -130,6 +130,14 @@ def test_union_with_two_annotated_values_list(self):
         ).values_list('num', 'count')
         self.assertCountEqual(qs1.union(qs2), [(1, 0), (2, 1)])
 
+    def test_union_with_values_list_on_annotated_and_not_annotated(self):
```
Comment (timgraham) [inline tests/queries/test_qs_combinators.py]: not_annotated -> unannotated

**Comment by carltongibson** [comment]
This was on my list for this morning. I wake up and it's all resolved. Well done everybody. 👍

**Comment by felixxm** [comment]
@timgraham Thanks for the review, and for pointing out the lack of unpacking generalizations support in the older Python versions (I would forget about it :bowing_man:).

**Pull request merged**